pkg/resource/deploy/plan_executor.go

```diff
@@ -326,6 +355,23 @@ func (pe *planExecutor) refresh(callerCtx context.Context, opts Options, preview
 	stepExec.SignalCompletion()
 	stepExec.WaitForCompletion()
 
+	pe.rebuildDependencyGraph(resourceToStep, true /*refresh*/)
+
+	// NOTE: we use the presence of an error in the caller context in order to distinguish caller-initiated
+	// cancellation from internally-initiated cancellation.
+	canceled := callerCtx.Err() != nil
+
+	if stepExec.Errored() {
+		pe.reportExecResult("failed", preview)
+		return result.Bail()
+	} else if canceled {
+		pe.reportExecResult("canceled", preview)
+		return result.Bail()
+	}
+	return nil
+}
+
+func (pe *planExecutor) rebuildDependencyGraph(resourceToStep map[*resource.State]Step, refresh bool) {
```

Comment:
Strictly speaking this is also rebuilding the set of old resources. Should we just call this `rebuildBaseState` or similar?

Reply:
Done.